crates/iceberg/src/writer/file_writer/clustered_writer.rs

```diff
@@ -0,0 +1,155 @@
+use crate::{Error, ErrorKind, Result};
+
+/// todo doc
+#[derive(Clone)]
+pub struct ClusteredWriterBuilder<B: FileWriterBuilder> {
+    inner_builder: B,
+    partition_spec: PartitionSpecRef,
+    table_schema: SchemaRef,
+    arrow_schema: ArrowSchemaRef,
+}
+
+impl<B: FileWriterBuilder> ClusteredWriterBuilder<B> {
+    #[allow(dead_code)]
+    pub fn new(
+        inner_builder: B,
+        partition_spec: PartitionSpecRef,
+        table_schema: SchemaRef,
+        arrow_schema: ArrowSchemaRef,
+    ) -> Self {
+        Self {
+            inner_builder,
+            partition_spec,
+            table_schema,
+            arrow_schema,
+        }
+    }
+}
+
+impl<B: FileWriterBuilder> FileWriterBuilder for ClusteredWriterBuilder<B> {
+    type R = ClusteredWriter<B>;
+
+    async fn build(self) -> Result<Self::R> {
+        let splitter = RecordBatchPartitionSplitter::new(
+            self.arrow_schema,
+            self.table_schema.clone(),
+            self.partition_spec.clone(),
+        )?;
+
+        Ok(ClusteredWriter {
+            inner: None,
+            inner_builder: self.inner_builder,
+            splitter,
+            table_schema: self.table_schema,
+            partition_spec: self.partition_spec,
+            current_partition_key: None,
+            data_file_builders: vec![],
+        })
+    }
+}
+
+pub struct ClusteredWriter<B: FileWriterBuilder> {
+    inner: Option<B::R>,
+    inner_builder: B,
+    splitter: RecordBatchPartitionSplitter,
+    table_schema: SchemaRef,
+    partition_spec: PartitionSpecRef,
+    current_partition_key: Option<PartitionKey>,
+    data_file_builders: Vec<DataFileBuilder>,
+}
+
+impl<B: FileWriterBuilder> FileWriter for ClusteredWriter<B> {
+    async fn write(&mut self, batch: &RecordBatch) -> Result<()> {
+        let splits = self.splitter.split(batch)?;
+        if splits.len() > 1 {
+            // todo revisit this, should we assume one batch can contain at most one partition's data?
+            return Err(Error::new(
+                ErrorKind::DataInvalid,
+                "Records from multiple partitions found in one record batch!",
+            ));
+        }
+
+        if let Some((partition_value, record_batch)) = splits.first() {
+            let partition_key = PartitionKey::new(
+                self.partition_spec.as_ref().clone(),
+                self.table_schema.clone(),
+                partition_value.clone(),
+            );
+
+            if self
+                .current_partition_key
+                .as_ref()
+                .is_none_or(|pk| pk != &partition_key)
+            {
+                if let Some(inner) = self.inner.take() {
+                    // Close the current writer, roll to a new file
+                    self.data_file_builders.extend(inner.close().await?);
+
+                    // Start a new writer
+                    // TODO how to pass partition key to inner builder??
```

Comment:
Hi @liurenjie1024 ,

I think we've discussed a while ago that we don't really need a new `PartitionWriter` trait, as using the decorator pattern should be enough, but I'm having second thought here: it's tricky to configure `PartitionKey` for the `inner_builder` without changing the `FileBuilder` trait.. I feel like we need to change at least the FileWriterBuilder trait:

```
trait FileWriterBuilder {
...
fn with_partition_key();
}
```

Would love to hear your thoughts!